File: backend/spec/factories/devis.rb
# frozen_string_literal: true

FactoryBot.define do
  factory :devis do
    organisation
    client { association(:client, organisation: organisation) }

    numero { nil }
    statut { "brouillon" }
    objet { "Devis de test" }
    date_emission { Date.current }
    date_validite { 30.days.from_now.to_date }
    total_ht { 0 }
    total_tva { 0 }
    total_ttc { 0 }
    conditions { nil }

    trait :avec_ligne do
      after(:create) do |devis|
        create(:ligne_devis, devis: devis, organisation: devis.organisation)
        devis.reload
      end
    end

    trait :envoye do
      after(:create) do |devis|
        create(:ligne_devis, devis: devis, organisation: devis.organisation) if devis.lignes_devis.empty?
        devis.reload

        numero = "DEV-#{Date.current.year}-#{format('%04d', rand(1000..9999))}"

        devis.update_columns(statut: "envoye", numero: numero, updated_at: Time.current)
      end
    end
  end
end

File: frontend/src/types/devis.ts
import type { Client } from "./client";
import type { Money } from "./facture";
import type { LigneDevis } from "./ligneDevis";

// Machine à états réelle (DevisStatutService::TRANSITIONS, backend) :
// brouillon -> envoye -> accepte / refuse. Pas de 5e statut "converti" — la
// conversion est un booléen DÉRIVÉ (cf. `converti` ci-dessous), jamais un
// statut stocké.
export type DevisStatut = "brouillon" | "envoye" | "accepte" | "refuse";

export type DevisFactureGeneree = {
  id: string;
  numero: string | null;
};

export type Devis = {
  id: string;
  client_id: string;
  client?: Pick<
    Client,
    | "id"
    | "raison_sociale"
    | "siret"
    | "identifiant_routage_pa"
    | "email"
    | "ville"
    | "pays"
    | "type"
  > | null;
  numero: string | null;
  objet: string | null;
  statut: DevisStatut;
  total_ht: Money;
  total_tva: Money;
  total_ttc: Money;
  conditions: string | null;
  date_emission: string | null;
  date_validite: string | null;
  // Dérivés côté backend (Devis#expire?/#converti?) — jamais stockés, jamais
  // recalculés côté front.
  expire: boolean;
  converti: boolean;
  facture_generee: DevisFactureGeneree | null;
  created_at?: string | null;
  updated_at?: string | null;
  lignes_count?: number;
  lignes_devis?: LigneDevis[];
};

export type DevisInput = {
  client_id: string;
  objet?: string | null;
  date_emission?: string | null;
  date_validite?: string | null;
  conditions?: string | null;
};

export type DevisUpdateInput = Partial<DevisInput>;
